Fix converting log level to string.

gilbertoabram opened the pull request:
Enabling this output causes the simulator to crash due to trying to access the log level as a pointer.

Is it an option to use the `format` attribute for `gpi_log_` when the compiler supports it to get warnings?

<!--

Thanks for improving cocotb! Here are some points to make this as smooth as possible.
Not all of them may be applicable.

Most important: please explain *why* you are proposing this change.

* Make sure you have read https://github.com/cocotb/cocotb/blob/master/CONTRIBUTING.md
* Extend or add a test under `tests/test_cases/`.
* Add documentation under `docs/source/`,
  docstrings in Python code, or Doxygen markup in C/C++ code.
  Use ``versionadded``/``versionchanged``/``deprecated``.
* Add a newsfragment - see `docs/source/newsfragments/README.rst`.
* Use `closes #XXXX` to auto-close the issue that this PR fixes (if such).

-->

ktbarrett reviewed (approved):
Whoops. ~~The danger of poorly named functions.~~ Apparently it's just a cast? I figured the compiler would see that and complain. Maybe we need something to get printf-style function checking to work.

ktbarrett commented:
Could this be https://github.com/cocotb/cocotb/issues/4526?

gilbertoabram commented:
> Could this be #4526?

That does sound very similar indeed. One difference is that in issue_348 the clock is implemented via Timer, so I believe it should also be reproducible with just two timers (at least with my setup and for some tests).

gilbertoabram commented:
Allright, so at least in my case, cocotb correctly calls `vpi_remove_cb` for the second timer, and it succeeds (returns 1), so we delete the cb obj. but later on verilator still calls the callback :(

```
-V{t1,1561}- vpi: callCbs reason=9
-V{t1,1562}- vpi: reason_callback reason=9 id=4
HANDLING: 0x555555bd6610
REMOVED: 0x55555584f540
-V{t1,1563}- vpi: vpi_remove_cb 0x555555a42b28
REMOVED ERR: 1
REMOVED: 0x555555a44d80
-V{t1,1564}- vpi: vpi_remove_cb 0x555555ad6128
REMOVED ERR: 1
   100.00ns INFO     cocotb.regression                  issue_348.issue_348_rising passed
   100.00ns INFO     cocotb.regression                  running issue_348.issue_348_falling (2/3)
                                                            Start two monitors on FallingEdge
-V{t1,1565}- vpi: vpi_register_cb reason=9 id=56 time=100001 obj=(nil)
REGISTERED: 0x555555c59e20
-V{t1,1566}- vpi: vpi_remove_cb 0x555555aef8c8
-V{t1,1567}- vpi: reason_callback reason=9 id=53
HANDLING: 0x555555a44d80
```
(I added some printouts, 0x555555a44d80 is the offending cb obj)

one fix would be to set `m_removed` as mentioned in #4526, however that would need to be done only if it was about to fire, otherwise we'll be leaking the cb objects if they never get called back.

ktbarrett commented:
Well this isn't breaking anything, so after #4536 comes in which adds "may-fail" to our CI, this can come in.

cmarqu commented:
> after #4536 comes in which adds "may-fail" to our CI, this can come in.

#4536 is closed unmerged?

ktbarrett commented:
#4529 is replacement. Apologize for all the shuffling.
